README.md

# animated_button_bar  
  
AnimatedButtonBar is a flutter library that allows you to create a row of buttons with an animation on selection.  
  
##  Getting started  
  
Add this to your app's `pubspec.yaml` file:  
```yaml  
dependencies:  
  animated_button_bar: ^0.0.2
```  
  
## Usage  
  
Then you have to import the package with:  
  
```dart  
import 'package:animated_button_bar/animated_button_bar.dart';  
```  
  
And use an `AnimatedBarButton` like this:  
```dart  
  Column(  
          mainAxisAlignment: MainAxisAlignment.center,  
          children: [  
            //inverted selection button bar  
            AnimatedButtonBar(  
              radius: 8.0,  
              padding: const EdgeInsets.all(16.0),  
              invertedSelection: true,  
              children: [  
                ButtonBarEntry(onTap: () => print('First item tapped'), child: Text('Day')),  
                ButtonBarEntry(onTap: () => print('Second item tapped'), child: Text('Week')),  
                ButtonBarEntry(onTap: () => print('Third item tapped'), child: Text('Month')),  
                ButtonBarEntry(onTap: () => print('Fourth item tapped'), child: Text('Year'))  
              ],  
            ),  
            //You can populate it with different types of widgets like Icon  
            AnimatedButtonBar(  
              radius: 32.0,  
              padding: const EdgeInsets.all(16.0),  
              backgroundColor: Colors.blueGrey.shade800,  
              foregroundColor: Colors.blueGrey.shade300,  
              elevation: 24,  
              borderColor: Colors.white,  
              borderWidth: 2,  
              innerVerticalPadding: 16,  
              children: [  
                ButtonBarEntry(onTap: () => print('First item tapped'), child: Icon(Icons.person)),  
                ButtonBarEntry(onTap: () => print('Second item tapped'), child: Icon(Icons.people)),  
              ],  
            ),  
          ],  
        ),  
```  
  
## Result  
  
![](example.gif)  
  
## parameters  
| parameter                  | description                                                                           | default                                                                                                                                                                               |  
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |  
| animationDuration          | animation duration when tap                                            | const Duration(milliseconds: 200)                                                                                                                                                    |  
| backgroundColor                       | main color of the widget                                                                   | Theme.of(context).backgroundColor                                                                                                                                                                                  |  
| borderColor                | color of the border                                                                 | null                                                                                                                                                                            |  
| borderWidth                | width of the border                                                                 | borderColor != null ? 1.0 : 0.0;                                                                                                                                                                          |  
| children                | A list of ButtonBarEntry to display                                                                 | required                                                                                                                                                                          |  
| curve                | the curve for the animation                                                                      | Curves.fastOutSlowIn                                                                                                        |  
| elevation                | double                                                                     | 0                                                                                                        |  
| foregroundColor                | color for the selection                                                                    | Theme.of(context).accentColor    
| invertedSelection                      | bool           |    false                                                                                                                                                                                   |                                                                                                     |  
| radius                      | double           |    0.0                                                                                                                                                                                   |  
| verticalPadding                    | double | 8.0                                                                                                                                                                                 |

CHANGELOG.md
## [1.0.0] - 2023-02-18

* Added AnimatedButtonController to set index programmatically

## [0.0.2+1] - 2021-04-06

* Dart code formatting

## [0.0.2] - 2021-04-06

* Null safety
* Inverted selection feature

## [0.0.1] - 2021-03-29

* First release


lib/animated_button_bar.dart
library animated_button_bar;

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

///A row of buttons with animated selection
class AnimatedButtonBar extends StatefulWidget {
  ///Duration for the selection animation
  final Duration animationDuration;
  final Color? backgroundColor;
  final Color? foregroundColor;
  final double radius;

  ///A list of [ButtonBarEntry] to display
  final List<ButtonBarEntry> children;
  final double innerVerticalPadding;
  final double elevation;
  final Color? borderColor;
  final double? borderWidth;
  final Curve curve;
  final EdgeInsets padding;

  ///Invert color of the child when true
  final bool invertedSelection;

  ///Allows to programatically interact with widget
  final AnimatedButtonController? controller;

  const AnimatedButtonBar({
    Key? key,
    required this.children,
    this.animationDuration = const Duration(milliseconds: 200),
    this.backgroundColor,
    this.foregroundColor,
    this.radius = 0.0,
    this.innerVerticalPadding = 8.0,
    this.elevation = 0,
    this.borderColor,
    this.controller,
    this.borderWidth,
    this.curve = Curves.fastOutSlowIn,
    this.padding = const EdgeInsets.all(0),
    this.invertedSelection = false,
  }) : super(key: key);

  @override
  _AnimatedButtonBarState createState() => _AnimatedButtonBarState();
}

class _AnimatedButtonBarState extends State<AnimatedButtonBar> {

  late AnimatedButtonController _controller;
  @override
  void initState() {
    _controller = widget.controller ?? AnimatedButtonController();
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    Color backgroundColor =
        widget.backgroundColor ?? Theme.of(context).colorScheme.background;
    return ChangeNotifierProvider(
      create: (context) => _controller,
      child: Consumer<AnimatedButtonController>(
        builder: (context, animatedButton, child) {
          return Padding(
            padding: widget.padding,
            child: LayoutBuilder(
                builder: (BuildContext context, BoxConstraints constraints) {
              return Card(
                color: backgroundColor,
                shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.all(Radius.circular(widget.radius)),
                    side: BorderSide(
                      color: widget.borderColor ?? Colors.transparent,
                      width: widget.borderWidth ??
                          (widget.borderColor != null ? 1.0 : 0.0),
                    )),
                elevation: widget.elevation,
                child: Stack(
                  fit: StackFit.loose,
                  children: [
                    AnimatedPositioned(
                      top: 0,
                      bottom: 0,
                      left: constraints.maxWidth / widget.children.length * animatedButton.index,
                      right: (constraints.maxWidth / widget.children.length) *
                          (widget.children.length - animatedButton.index - 1),
                      duration: widget.animationDuration,
                      curve: widget.curve,
                      child: Container(
                        decoration: BoxDecoration(
                          color:
                              widget.foregroundColor ?? Theme.of(context).colorScheme.secondary,
                          borderRadius:
                              BorderRadius.all(Radius.circular(widget.radius)),
                        ),
                      ),
                    ),
                    Row(
                      children: widget.children
                          .asMap()
                          .map((i, sideButton) => MapEntry(
                                i,
                                Expanded(
                                  child: InkWell(
                                    onTap: () {
                                      try {
                                        sideButton.onTap();
                                      } catch (e) {
                                        print('onTap implementation is missing');
                                      }
                                      animatedButton.setIndex(i);
                                    },
                                    borderRadius: BorderRadius.all(
                                        Radius.circular(widget.radius)),
                                    child: Padding(
                                      padding: EdgeInsets.symmetric(
                                          vertical: widget.innerVerticalPadding),
                                      child: Center(
                                          child: ColorFiltered(
                                              colorFilter: ColorFilter.mode(
                                                  backgroundColor,
                                                  widget.invertedSelection &&
                                                      animatedButton.index == i
                                                      ? BlendMode.srcIn
                                                      : BlendMode.dstIn),
                                              child: sideButton.child)),
                                    ),
                                  ),
                                ),
                              ))
                          .values
                          .toList(),
                    ),
                  ],
                ),
              );
            }),
          );
        }
      ),
    );
  }
}

class ButtonBarEntry {
  final Widget child;
  final VoidCallback onTap;
  ButtonBarEntry({required this.child, required this.onTap});
}

///controller for AnimatedButtonBar widget
class AnimatedButtonController extends ChangeNotifier {

  int index = 0;

  ///change the index programmatically
  void setIndex(int requestedIndex) {
    index = requestedIndex;
    notifyListeners();
  }
}

example/lib/main.dart
import 'package:animated_button_bar/animated_button_bar.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'AnimatedButtonBar',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(primarySwatch: Colors.blue, brightness: Brightness.dark),
      home: SampleView(),
    );
  }
}

class SampleView extends StatefulWidget {
  @override
  _SampleViewState createState() => _SampleViewState();
}

class _SampleViewState extends State<SampleView> {
  AnimatedButtonController _controller = AnimatedButtonController();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.blueGrey.shade900,
      body: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          //inverted selection button bar
          AnimatedButtonBar(
            radius: 8.0,
            padding: const EdgeInsets.all(16.0),
            invertedSelection: true,
            children: [
              ButtonBarEntry(
                  onTap: () => print('First item tapped'), child: Text('Day')),
              ButtonBarEntry(
                  onTap: () => print('Second item tapped'),
                  child: Text('Week')),
              ButtonBarEntry(
                  onTap: () => print('Third item tapped'),
                  child: Text('Month')),
              ButtonBarEntry(
                  onTap: () => print('Fourth item tapped'), child: Text('Year'))
            ],
          ),
          //You can populate it with different types of widgets like Icon
          AnimatedButtonBar(
            radius: 32.0,
            padding: const EdgeInsets.all(16.0),
            backgroundColor: Colors.blueGrey.shade800,
            foregroundColor: Colors.blueGrey.shade300,
            elevation: 24,
            borderColor: Colors.white,
            borderWidth: 2,
            innerVerticalPadding: 16,
            children: [
              ButtonBarEntry(
                  onTap: () => print('First item tapped'),
                  child: Icon(Icons.person)),
              ButtonBarEntry(
                  onTap: () => print('Second item tapped'),
                  child: Icon(Icons.people)),
            ],
          ),
          AnimatedButtonBar(
            controller: _controller,
            radius: 32.0,
            padding: const EdgeInsets.all(16.0),
            backgroundColor: Colors.blueGrey.shade800,
            foregroundColor: Colors.blueGrey.shade300,
            elevation: 24,
            borderColor: Colors.white,
            borderWidth: 2,
            innerVerticalPadding: 16,
            children: [
              ButtonBarEntry(
                  onTap: () => print('First item tapped'),
                  child: Icon(Icons.sunny)),
              ButtonBarEntry(
                  onTap: () => print('Second item tapped'),
                  child: Icon(Icons.nightlight_round)),
            ],
          ),
          ElevatedButton(
              onPressed: () {
                _controller.setIndex(_controller.index == 0 ? 1 : 0);
              },
              child: Text('Set index'))
        ],
      ),
    );
  }
}
